lib/machines/carousel.machine.ts

```diff
@@ -0,0 +1,97 @@
+import { assign, createMachine } from "xstate";
+
+export interface CarouselMachineContext {
+  activeIndex: number;
+  totalSlideCount: number;
+  interval: number;
+};
+
+export type CarouselMachineEvent =
+| { type: "UPDATE_INDEX"; index?: number }
+| { type: "PAUSE" }
+| { type: "RESET" }
+| { type: "AUTO_PLAY" };
+
+const createCarouselMachine = (initialContext: CarouselMachineContext = {
+  activeIndex: 0, totalSlideCount: 5, interval: 4000
+}) => createMachine<
+  CarouselMachineContext,
+  CarouselMachineEvent
+>(
+  {
+    id: "carouselMachine",
+    initial: "idle",
+    context: initialContext,
+    states: {
+      idle: {
+        on: {
+          UPDATE_INDEX: {
+            actions: ["updateActiveIndex"],
+          },
+          AUTO_PLAY: "playing",
+        },
+      },
+      paused: {
+        on: {
+          AUTO_PLAY: "playing",
+          RESET: {
+            target: "idle",
+            actions: ["resetActiveIndex"],
+          },
+          UPDATE_INDEX: {
+            actions: ["updateActiveIndex"],
+          },
+        },
+      },
+      playing: {
+        invoke: {
+          src: "autoUpdateActiveIndex",
+        },
+        on: {
+          RESET: {
+            target: "idle",
+            actions: ["resetActiveIndex"],
+          },
+          PAUSE: "paused",
+          UPDATE_INDEX: {
+            actions: ["updateActiveIndex"],
+          },
+        },
+      },
+    },
+  },
+  {
+    actions: {
+      resetActiveIndex: assign({
+        activeIndex: 0,
+      }),
+      updateActiveIndex: assign({
+        activeIndex: (context, event) => {
+          if (event.type === "UPDATE_INDEX" && event.index) {
+            return event.index
+          }
+
+          if (context.activeIndex === context.totalSlideCount) {
+            return 0
+          } else {
+            return context.activeIndex + 1
+          }
+        } 
+      }),
+
+    },
+    services: {
+      autoUpdateActiveIndex: (context) => (cb) => {
+        const interval = setInterval(() => cb({ type: "UPDATE_INDEX" }), context.interval);
+
+        return () => {
+          clearInterval(interval);
+        };
+      },
+    },
+  }
+);
+
+const carouselMachine = createCarouselMachine()
+
+export default carouselMachine;
```

Comment:
There might be something we can do here to mean you could just export the factory that creates the machine! This might be a little confusing to read.

Reply:
For sure! Do you mean something like...

`export default createCarouselMachine();`

Reply:
I actually meant something at the lower level, so you could do:

`export default createCarouselMachine`

Reply:
We get a runtime error as `pages/machines/[id].tsx` is looking for the default export.

How about 

```javascript 
export {
  createCarouselMachine
};

export default createCarouselMachine();

Reply:
I was thinking that maybe `pages/machines/[id].tsx` could _either_ accept a factory function as the default export, or a normal machine. Factory functions are very useful but largely undocumented as a pattern, makes sense that we should have more examples of them. `export default createCarouselMachine();` will work great for now, we can revisit this later.

Reply:
Thanks, yeah that makes sense 👍